File: apps/docs/src/prework/examples/python/states.py
from dataclasses import dataclass
from typing import assert_never


@dataclass(frozen=True)
class Draft:
    id: str


@dataclass(frozen=True)
class Approved:
    id: str
    approved_by: str


Permit = Draft | Approved


# 状態の形だけを学ぶ例。業務の承認条件は省略する。
def approve(draft: Draft, approved_by: str) -> Approved:
    return Approved(draft.id, approved_by)


def label(permit: Permit) -> str:
    if isinstance(permit, Draft):
        return "申請中"
    if isinstance(permit, Approved):
        return f"承認者: {permit.approved_by}"
    assert_never(permit)

File: apps/docs/src/prework/examples/rust/states.rs
pub struct Draft {
    pub id: String,
}

pub struct Approved {
    pub id: String,
    pub approved_by: String,
}

pub enum Permit {
    Draft(Draft),
    Approved(Approved),
}

// 状態の形だけを学ぶ例。業務の承認条件は省略する。
// 比較しやすいように借用し、元のDraftを残す。
pub fn approve(draft: &Draft, approved_by: String) -> Approved {
    Approved {
        id: draft.id.clone(),
        approved_by,
    }
}

pub fn label(permit: &Permit) -> String {
    match permit {
        Permit::Draft(_) => "申請中".into(),
        Permit::Approved(approved) => format!("承認者: {}", approved.approved_by),
    }
}
